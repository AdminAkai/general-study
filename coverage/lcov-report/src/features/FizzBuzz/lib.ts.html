
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/features/FizzBuzz/lib.ts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> / <a href="index.html">src/features/FizzBuzz</a> lib.ts</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>0/1</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>0/0</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>0/0</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>0/1</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line low'></div>
    <pre><table class="coverage">
<tr><td class="line-count quiet"><a name='L1'></a><a href='#L1'>1</a>
<a name='L2'></a><a href='#L2'>2</a>
<a name='L3'></a><a href='#L3'>3</a>
<a name='L4'></a><a href='#L4'>4</a>
<a name='L5'></a><a href='#L5'>5</a>
<a name='L6'></a><a href='#L6'>6</a>
<a name='L7'></a><a href='#L7'>7</a>
<a name='L8'></a><a href='#L8'>8</a>
<a name='L9'></a><a href='#L9'>9</a>
<a name='L10'></a><a href='#L10'>10</a>
<a name='L11'></a><a href='#L11'>11</a>
<a name='L12'></a><a href='#L12'>12</a>
<a name='L13'></a><a href='#L13'>13</a>
<a name='L14'></a><a href='#L14'>14</a>
<a name='L15'></a><a href='#L15'>15</a>
<a name='L16'></a><a href='#L16'>16</a>
<a name='L17'></a><a href='#L17'>17</a>
<a name='L18'></a><a href='#L18'>18</a>
<a name='L19'></a><a href='#L19'>19</a>
<a name='L20'></a><a href='#L20'>20</a>
<a name='L21'></a><a href='#L21'>21</a>
<a name='L22'></a><a href='#L22'>22</a>
<a name='L23'></a><a href='#L23'>23</a>
<a name='L24'></a><a href='#L24'>24</a>
<a name='L25'></a><a href='#L25'>25</a>
<a name='L26'></a><a href='#L26'>26</a>
<a name='L27'></a><a href='#L27'>27</a>
<a name='L28'></a><a href='#L28'>28</a>
<a name='L29'></a><a href='#L29'>29</a>
<a name='L30'></a><a href='#L30'>30</a>
<a name='L31'></a><a href='#L31'>31</a>
<a name='L32'></a><a href='#L32'>32</a>
<a name='L33'></a><a href='#L33'>33</a>
<a name='L34'></a><a href='#L34'>34</a>
<a name='L35'></a><a href='#L35'>35</a>
<a name='L36'></a><a href='#L36'>36</a>
<a name='L37'></a><a href='#L37'>37</a>
<a name='L38'></a><a href='#L38'>38</a>
<a name='L39'></a><a href='#L39'>39</a>
<a name='L40'></a><a href='#L40'>40</a>
<a name='L41'></a><a href='#L41'>41</a>
<a name='L42'></a><a href='#L42'>42</a>
<a name='L43'></a><a href='#L43'>43</a>
<a name='L44'></a><a href='#L44'>44</a>
<a name='L45'></a><a href='#L45'>45</a>
<a name='L46'></a><a href='#L46'>46</a>
<a name='L47'></a><a href='#L47'>47</a>
<a name='L48'></a><a href='#L48'>48</a>
<a name='L49'></a><a href='#L49'>49</a>
<a name='L50'></a><a href='#L50'>50</a>
<a name='L51'></a><a href='#L51'>51</a>
<a name='L52'></a><a href='#L52'>52</a>
<a name='L53'></a><a href='#L53'>53</a>
<a name='L54'></a><a href='#L54'>54</a>
<a name='L55'></a><a href='#L55'>55</a>
<a name='L56'></a><a href='#L56'>56</a>
<a name='L57'></a><a href='#L57'>57</a>
<a name='L58'></a><a href='#L58'>58</a>
<a name='L59'></a><a href='#L59'>59</a>
<a name='L60'></a><a href='#L60'>60</a>
<a name='L61'></a><a href='#L61'>61</a>
<a name='L62'></a><a href='#L62'>62</a>
<a name='L63'></a><a href='#L63'>63</a>
<a name='L64'></a><a href='#L64'>64</a>
<a name='L65'></a><a href='#L65'>65</a>
<a name='L66'></a><a href='#L66'>66</a>
<a name='L67'></a><a href='#L67'>67</a>
<a name='L68'></a><a href='#L68'>68</a>
<a name='L69'></a><a href='#L69'>69</a>
<a name='L70'></a><a href='#L70'>70</a>
<a name='L71'></a><a href='#L71'>71</a>
<a name='L72'></a><a href='#L72'>72</a>
<a name='L73'></a><a href='#L73'>73</a>
<a name='L74'></a><a href='#L74'>74</a>
<a name='L75'></a><a href='#L75'>75</a>
<a name='L76'></a><a href='#L76'>76</a>
<a name='L77'></a><a href='#L77'>77</a>
<a name='L78'></a><a href='#L78'>78</a>
<a name='L79'></a><a href='#L79'>79</a>
<a name='L80'></a><a href='#L80'>80</a>
<a name='L81'></a><a href='#L81'>81</a>
<a name='L82'></a><a href='#L82'>82</a>
<a name='L83'></a><a href='#L83'>83</a>
<a name='L84'></a><a href='#L84'>84</a>
<a name='L85'></a><a href='#L85'>85</a>
<a name='L86'></a><a href='#L86'>86</a>
<a name='L87'></a><a href='#L87'>87</a>
<a name='L88'></a><a href='#L88'>88</a></td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">interface QuestionAnswer {
  question: string
  answer: string
}
&nbsp;
export const <span class="cstat-no" title="statement not covered" >QuestionsAnswers: QuestionAnswer[] = [</span>
  {
    question: "What's a Promise in Javascript",
    answer:
      "Javascript's way of handling asynchronous operations, represented by an object that is either fulfilled or rejected method.",
  },
  {
    question: 'When would you choose async/await vs then/catch approach',
    answer:
      'You would choose async/await for both versatility and readability. It is much easier to both read and run an async operation and await for the result in a variable rather than be launched into callback hell with the .then approach.',
  },
  {
    question: 'What if an error is thrown during the async/await',
    answer:
      'If an error is thrown in an async/await function then it short circuits unless handled in a try/catch block',
  },
  {
    question:
      'What is type coercion and how does this impact the == and === operator behavior',
    answer:
      "Type coercion is when Javascript changes the data type of a value in order to be able to use them in an operation. This means 2 == '2' will be true as they will both be type casted into number, while 2 === '2' will be false as it is a strict equality operation which also evaluates the type.",
  },
  {
    question: 'What is the DOM and how does it relate to React',
    answer:
      'The DOM is the document object model, and a way for Javascript to interact with the webpage and allows it to be manipulated, without it React would not exist',
  },
  {
    question: 'What about the Virtual DOM and why is it useful',
    answer:
      "The Virtual DOM is React's way of rehydrating the HTML on the web page with new data from state or async calls, and React updates the real DOM based on the diff between the real DOM and the virtual DOM",
  },
  {
    question: 'Whats the lifecycle of a React Component',
    answer: 'The lifecycle is mounting, updating, and unmounting',
  },
  {
    question:
      'What would it look like in a React component to make an API call and render the results',
    answer:
      'You would use a hook to handle the side effect of making an API call and render the data in the return/jsx of the function',
  },
  {
    question: 'What are custom hooks and when would you use them',
    answer:
      'Custom hooks are basically React functions that have an internal state and useEffect and return that state rather than return a component, you would use them whenever you want to abstract a side effect to be re-used in other React components',
  },
  {
    question:
      'If you had to keep state across the entire app, how would you do this?',
    answer:
      'I would use Redux to maintain a global state tree, but an alternative solution would be React Context API which is what React Redux uses under the hood',
  },
  {
    question: 'When would you use Redux vs Context API',
    answer:
      "Personally for me, all the time, as long as the project is architectured well then Redux's state abstraction is a lot more readable and easier to scale than making a million Contexts",
  },
  {
    question:
      'How is uniqueness determined when rendering multiple elements with a for loop',
    answer: 'Key property on the parent tag of that element',
  },
  {
    question:
      'How do you prevent unnecessary rendering of a component when given the same input',
    answer:
      'Memoize the component with memo(), if it only affects a certain computation or function though then useMemo or useCallback will also suffice, and if you ONLY need the component to re-render based on specific props then it can be added to the dependency array of a useEffect hook',
  },
  {
    question:
      'If we had to listen to a number of mouse clicks inside a React component, how would we do that efficiently?',
    answer:
      'If it was just general mouse clicks and not anything specific to a React Node then I would assign an event listener in a useEffect hook and clean it up in the return',
  },
  {
    question:
      'Why is it important to clean up the event listener you mentioned',
    answer:
      "Its important because when the component unmounts the listener remains as it is attached to the window, so you want to avoid memory leaks by cleaning it up, but also because on every re-render it'll attach a new event listener to the window.",
  },
]
&nbsp;</pre></td></tr></table></pre>

                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2023-12-09T19:50:42.390Z
            </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    