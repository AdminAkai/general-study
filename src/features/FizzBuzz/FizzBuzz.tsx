import { ChangeEventHandler, FC, useState } from 'react'

import {
  FizzBuzzContainer,
  FizzBuzzDisplay,
  FizzBuzzHeader,
  FizzBuzzInput,
  FizzBuzzSubmit,
} from './styledComponents'

const FizzBuzz: FC = () => {
  const [input, setInput] = useState<string>('')
  const [fizzBuzzValue, setFizzBuzzValue] = useState<string>('')

  const handleChange: ChangeEventHandler<HTMLInputElement> = (e) => {
    setInput(e.target.value)
  }

  const handleFizzBuzz = () => {
    const compare = parseInt(input)
    const array: Array<string | number> = []

    for (let i = 0; i <= compare; i++) {
      if (i === 0) {
        array.push(i)
        continue
      }
      if (i % 3 === 0) {
        array.push('Fizz')
        continue
      }
      if (i % 5 === 0) {
        array.push('Buzz')
        continue
      }
      if (i % 5 === 0 && i % 5 === 0) {
        array.push('FizzBuzz')
        continue
      }
      array.push(i)
    }

    setFizzBuzzValue(array.join(' '))
  }

  return (
    <FizzBuzzContainer>
      <FizzBuzzHeader>Enter a number to FizzBuzz here!</FizzBuzzHeader>
      <FizzBuzzInput
        aria-label='FizzBuzz Input'
        onChange={handleChange}
        type='number'
        value={input}
      />
      <FizzBuzzSubmit onClick={handleFizzBuzz}>FizzBuzz</FizzBuzzSubmit>
      <FizzBuzzDisplay aria-label='FizzBuzz Display'>
        The FizzBuzz from {input || 'input'}: {fizzBuzzValue}
      </FizzBuzzDisplay>
    </FizzBuzzContainer>
  )
}

export default FizzBuzz
