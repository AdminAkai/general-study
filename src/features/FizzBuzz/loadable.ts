/**
 * Asynchronously loads the component for container
 */

import loadable from 'src/shared/utils/loadable'

export default loadable(() => import('./FizzBuzz'), {
  fallback: null,
})
