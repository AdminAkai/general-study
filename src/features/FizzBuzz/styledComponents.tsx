import styled from 'styled-components'

export const FizzBuzzContainer = styled.div`
  width: 100%;
  height: 100%;
  padding: 0 var(--padding-huge);
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: var(--spacing-extra-large);

  @media screen and (max-width: ${({ theme }) => theme.mobileMediaQuery}) {
    height: 100vh;
    padding: 1.5rem;
    flex-direction: column;
    justify-content: flex-start;
  }
`

export const FizzBuzzInput = styled.input`
  border-radius: 4px;
`

export const FizzBuzzSubmit = styled.button`
  border-radius: 4px;
  background: none;
  color: ${({ theme }) => theme.Adminis};
  border: 2px solid ${({ theme }) => theme.Adminis};
  padding: 8px 16px;
  outline: inherit;
  cursor: pointer;
  transition: background-color 200ms ease;

  &:hover {
    background-color: ${({ theme }) => theme.Adminis};
    color: ${({ theme }) => theme.primary};
  }
`
export const FizzBuzzHeader = styled.h1`
  font-size: 2rem;
  color: ${({ theme }) => theme.Adminis};
  padding: 0;
  margin: 0;
`

export const FizzBuzzDisplay = styled.div`
  text-align: center;
  font-size: 2rem;
  color: ${({ theme }) => theme.Adminis};
`
