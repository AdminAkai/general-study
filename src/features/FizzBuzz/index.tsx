import FizzBuzz from './loadable'

export default FizzBuzz
