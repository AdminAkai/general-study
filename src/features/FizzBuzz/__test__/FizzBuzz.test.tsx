import { render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'

import FizzBuzz from '../FizzBuzz.tsx'

describe('FizzBuzz tests', () => {
  it("Make sure FizzBuzz input FizzBuzz's", async () => {
    render(<FizzBuzz />)
    const input = screen.getByLabelText('FizzBuzz Input')
    await userEvent.type(input, '5')
    expect(input).toHaveValue(5)
    const displayText = screen.getByLabelText('FizzBuzz Display')
    const fizzBuzzButton = screen.getByRole('button', { name: 'FizzBuzz' })
    await userEvent.click(fizzBuzzButton)
    expect(displayText).toHaveTextContent(
      'The FizzBuzz from 5: 0 1 2 Fizz 4 Buzz'
    )
  })

  it('Navbar snapshot renders correctly', () => {
    const tree = renderer.create(<FizzBuzz />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
