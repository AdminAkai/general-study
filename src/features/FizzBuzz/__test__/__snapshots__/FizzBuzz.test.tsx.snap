// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FizzBuzz tests Navbar snapshot renders correctly 1`] = `
<div
  className="sc-beyTiQ iIkXX"
>
  <h1
    className="sc-hLQTFJ clRMBF"
  >
    Enter a number to FizzBuzz here!
  </h1>
  <input
    aria-label="FizzBuzz Input"
    className="sc-guDMob fvXIsm"
    onChange={[Function]}
    type="number"
    value=""
  />
  <button
    className="sc-dmyDmy fFvcGr"
    onClick={[Function]}
  >
    FizzBuzz
  </button>
  <div
    aria-label="FizzBuzz Display"
    className="sc-eDLKEg esEbxC"
  >
    The FizzBuzz from 
    input
    : 
  </div>
</div>
`;
